src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/DataAccess.cs

```diff
@@ -52,25 +53,10 @@ public async Task SetupAccount(Account account)
             await dbConnection.InsertAsync(account);
         }
 
-        // Use this method to add `Status` column data as string as dapper would insert numeric enum value
-        public async Task InsertApplication(IncentiveApplication application)
+        public async Task<long> InsertWithEnumAsString<T>(T entity) where T : class
         {
             await using var dbConnection = new SqlConnection(_connectionString);
-            await dbConnection.ExecuteAsync(
-                "insert into IncentiveApplication(id, accountId, accountLegalEntityId, dateCreated, status, dateSubmitted, submittedByEmail, submittedByName) values " +
-                "(@id, @accountId, @accountLegalEntityId, @dateCreated, @status, @dateSubmitted, @submittedByEmail, @submittedByName)",
-                new
-                {
-                    application.Id,
-                    application.AccountId,
-                    application.AccountLegalEntityId,
-                    application.DateCreated,
-                    Status = application.Status.ToString(),
-                    application.DateSubmitted,
-                    application.SubmittedByEmail,
-                    application.SubmittedByName
-                });
+            return await dbConnection.InsertWithEnumAsString(entity);
```

Comment:
Dapper has custom TypeMappers e.g.

```
   public class EarningsTypeHandler : SqlMapper.TypeHandler<EarningType>
    {        
        public override EarningType Parse(object value)
        {
            Enum.TryParse(value.ToString(), out EarningType earningType);

            return earningType;
        }

        public override void SetValue(IDbDataParameter parameter, EarningType value)
        {
            parameter.Value = value.ToString();
        }
    }
```

    

   and register with
    
    SqlMapper.AddTypeHandler(typeof(EarningType), new EarningsTypeHandler());
    
    Note doesn't work with Dapper Contrib but will work with SqlMapper.

Reply:
i've added a generic method which reads the the property attributes to work out whether it needs to insert enum as string or not. please take a look.

Reply:
Nice one.

Can you change the new method to match the InsertAsync  e.g.

```
public static Task<int> InsertAsync<T>(this IDbConnection connection, T entityToInsert, bool enumToString = false, IDbTransaction transaction = null, int? commandTimeout = null, ISqlAdapter sqlAdapter = null) where T : class
        {
            if (enumToString)
            {
                ...
            }
            else
            {
                return Dapper.Contrib.Extensions.SqlMapperExtensions.InsertAsync(connection, entityToInsert, transaction, commandTimeout, sqlAdapter);
            }
        }

```

Reply:
good idea!